File: Data Structures & Algorithms/permutation-string/submission-1.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        mp1, mp2 = {}, {}
        n1, n2 = len(s1), len(s2)
        for i in s1:
            mp1[i] = 1 if i not in mp1 else mp1[i] + 1
        left, right = 0, 0
        while left < n2:
            if s2[left] in mp1:
                right = left
                while right < n2 and s2[right] in mp1:
                    mp2[s2[right]] = 1 if s2[right] not in mp2 else mp2[s2[right]] + 1
                    if len(mp1) == len(mp2) and mp1 == mp2:
                        return True
                    if mp2[s2[right]] > mp1[s2[right]]:
                        break
                    right += 1
            mp2 = {}
            left += 1
        return False

File: Data Structures & Algorithms/find-k-closest-elements/submission-1.cpp
class Solution {
public:
    vector<int> findClosestElements(vector<int>& arr, int k, int x) {
        int N = static_cast<int>(arr.size());
        auto it = lower_bound(arr.begin(), arr.end(), x);
        deque<int> answer;
        if (it == arr.end()) {
            for (int i = N-k; i < N; i++) {
                answer.push_back(arr[i]);
            }
            return vector<int>(answer.begin(), answer.end());
        }
        if (it == arr.begin()) {
            for (int i = 0; i < k; i++) {
                answer.push_back(arr[i]);
            }
            return vector<int>(answer.begin(), answer.end());
        }
        int index = it - arr.begin();
        if (index -1 >= 0) {
            if (abs(arr[index-1] - x) <= abs(arr[index] - x)) {
                index --;
            }
        }
        int left = index, right = index;
        while (k --) {
            if (left < 0) {
                answer.push_back(arr[right]);
                right ++;
            } else if (right >= N) {
                answer.push_front(arr[left]);
                left --;
            } else {
                int left_diff = abs(arr[left] - x), right_diff = abs(arr[right] - x);
                if (left_diff <= right_diff) {
                    answer.push_front(arr[left]);
                    if (left == right) {
                        right ++;
                    }
                    left --;
                } else {
                    answer.push_back(arr[right]);
                    if (left == right) {
                        left --;
                    }
                    right ++;
                }
            }
        }
        return vector<int>(answer.begin(), answer.end());
    }
};